crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -121,35 +121,37 @@ use std::{
 ///
 /// [`SyncCell`]: bevy_utils::synccell::SyncCell
 /// [`Exclusive`]: https://doc.rust-lang.org/nightly/std/sync/struct.Exclusive.html
-pub trait SystemParam: Sized {
-    type State: SystemParamState;
-}
-
-pub type SystemParamItem<'w, 's, P> = <<P as SystemParam>::State as SystemParamState>::Item<'w, 's>;
-
-/// The state of a [`SystemParam`].
 ///
 /// # Safety
 ///
 /// It is the implementor's responsibility to ensure `system_meta` is populated with the _exact_
 /// [`World`] access used by the [`SystemParamState`].
 /// Additionally, it is the implementor's responsibility to ensure there is no
 /// conflicting access across all [`SystemParam`]'s.
-pub unsafe trait SystemParamState: Send + Sync + 'static {
-    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self;
+pub unsafe trait SystemParam: Sized {
+    type State: Send + Sync + 'static;
```

Comment:
I would like full doc comments on the methods and types here: now that the hierarchy is flatter this is actually possible to explain fairly directly.